hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -498,53 +498,63 @@ int SaraNcpClient::getSignalQuality(CellularSignalQuality* qual) {
     CHECK(checkParser());
     CHECK(queryAndParseAtCops(qual));
 
+    // Min and max RSRP index values multiplied by 100
+    // Min: -19.5 and max: -3
+    const int min_rsrq_mul_by_100 = -1950;
+    const int max_rsrq_mul_by_100 = -300;
+
     if (ncpId() == PLATFORM_NCP_SARA_R410) {
-        int rxlev, rxqual, rscp, ecn0, rsrq, rsrp;
-        auto resp = parser_.sendCommand("AT+CESQ");
-        int r = CHECK_PARSER(resp.scanf("+CESQ: %d,%d,%d,%d,%d,%d", &rxlev, &rxqual,
-                &rscp, &ecn0, &rsrq, &rsrp));
-        CHECK_TRUE(r == 6, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
-        r = CHECK_PARSER(resp.readResult());
-        CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
+        int rsrp;
+        int rsrq_n;
+        unsigned long rsrq_f;
 
-        switch (qual->strengthUnits()) {
-            case CellularStrengthUnits::RXLEV: {
-                qual->strength(rxlev);
-                break;
-            }
-            case CellularStrengthUnits::RSCP: {
-                qual->strength(rscp);
-                break;
-            }
-            case CellularStrengthUnits::RSRP: {
-                qual->strength(rsrp);
-                break;
+        // Set UCGED to mode 5 for RSRP/RSRQ values on R410M
+        CHECK_PARSER_OK(parser_.execCommand("AT+UCGED=5"));
+
+        // Default to invalid value because UCGED command does not explicitly give an ivalid value
+        qual->strength(255);
+        qual->quality(255);
+
+        auto resp = parser_.sendCommand("AT+UCGED?");
+        char buf[1024] = {};     // keep this string long in case there are multiple tower infos in the string
+        while (resp.hasNextLine()) {
+            CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
+            int r = 0;
+            if (strstr(buf, "+RSRP:")) {    // check for RSRP response
+                r = sscanf(buf, "\r\n+RSRP: %*d,%*d,\"%d.%*u\"",&rsrp);
+                CHECK_TRUE(r == 1, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
-            default: {
-                // Do nothing
-                break;
+            if (strstr(buf, "+RSRQ:")) {
+                r = sscanf(buf, "\r\n+RSRQ: %*d,%*d,\"%d.%lu\"",&rsrq_n, &rsrq_f);
+                CHECK_TRUE(r == 2, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
+            buf[0] = '\0';
         }
```

Comment:
So one thing that differs logically from the new code and old code, is that there is no check to make sure both RSRP and RSRQ are found.  Only IF they are found with `strstr` is there a CHECK_TRUE for a result.  So in theory RSRP OR RSRQ OR NEITHER could be found and the rest of the code will try to process the "found" values.  Previously if either was not found, the code would error out.  So we need some code here to keep this operation atomic and sane.

Reply:
Added code to check RSRP/Q are found before collecting results. Please see : 
https://github.com/particle-iot/device-os/blob/daf4878627829e849bf9c70e5ac46b25060f6a7a/hal/src/boron/network/sara_ncp_client.cpp#L525

I am parsing RSRP/Q separately so that even if one of the values is found, we will collect that. For the value that is not found it is defaulted to 255 in the beginning of UCGED command parsing. Please see here: https://github.com/particle-iot/device-os/blob/daf4878627829e849bf9c70e5ac46b25060f6a7a/hal/src/boron/network/sara_ncp_client.cpp#L512